.gitattributes

```diff
@@ -0,0 +1,5 @@
+# These files are text and should be normalized (convert crlf =&gt; lf)
+*.hpp text diff=cpp
```

Comment:
I haven't found any documentation for this option `text diff=cpp` (or replace `cpp` with `python` or `tex`). I did find this page, though: https://help.github.com/articles/dealing-with-line-endings/

Should we be using `text eol=lf`?

Reply:
It's in my first link: https://git-scm.com/docs/gitattributes

`lf` is default

Reply:
I'm a little weary of relying on defaults. Can we just set it? (I see why I didn't pick up that it was the default. The doc actually doesn't say it's the default. I guess you've picked up on it through this subsection? https://git-scm.com/docs/gitattributes#_checking_out_and_checking_in)

Any good reason to use `text diff=cpp` instead of just `text`?

Maybe I'll take a step back. Do you want the line endings to be `lf` in the repo, but whatever the OS does for the user? Or do you want the user to also check out as `lf`?

Reply:
Oh, sorry - Just want to make sure it's LF in the repository, don't actually care about working directory. So `eol=lf` will actually do in the working directory, I think.

`diff=cpp` I think just marks the filetype in the diff output, in case a user wants to set up their gitconfig to deal with those different types differently.

Reply:
wow... the doc is really tough to read. It's clear that to have the behavior you just described:
- this is wrong: `text eol=lf`
- this is also wrong: `text=auto` (this will leave crlf alone)
- I don't know what this does `text diff=cpp`. The doc doesn't say what it does for line endings.
- setting `text` alone marks it as a text file, but then it relies on the defaults (which may change and it isn't really doc'ed)

Anyway... if you think that's the best way to do it, I'll approve. Just FYI, it'd help me as a reviewer if you provided a little more of what you're trying to do + where to find it in the doc for this stuff.

Reply:
Good point. Also, text=auto will actually change line endings but it will just use Git's automatic textfile detection on files with that file extension. Setting it to text for that extension bypasses that detection. The implicit behavior seems to be to set LF as line endings in the repo for things marked text with either method, which is pretty crazy that it's not written explicitly anywhere. I don't think the default is likely to change between LF and CRLF given who maintains git, but it's crazy how this file works and how it's (not) documented.

From what I can tell, diff=cpp does nothing for line endings, it changes the diff chunk markings for files of that extension. It's described (poorly) in the section labeled "Defining a custom hunk-header." I think this just lets user settings to react to the hunk-headers.